constraintlayout/compose/src/main/java/androidx/constraintlayout/compose/MotionCarousel.kt

```diff
@@ -0,0 +1,396 @@
+ * on the right:
+ *
+ * start          [0] | [1] | [2]
+ *
+ * We can setup the backward state in the following way:
+ *
+ * backward           | [0] | [1] [3]
+ *
+ * And the forward state like:
+ *
+ * forward    [0] [1] | [2] |
+ *
+ * All three states together allowing to implement the Carousel motion we are looking for:
+ *
+ * backward           | [0] | [1] [3]
+ * start          [0] | [1] | [2]
+ * forward    [0] [1] | [2] |
+ *
+ * At the end of the swipe gesture, we instantly move back to the start state:
+ *
+ * start          [0] | [1] | [2]       -> gesture starts
+ * forward    [0] [1] | [2] |           -> gesture ends
+ * start          [0] | [1] | [2]       -> instant snap back to start state
+ *
+ * After the instant snap, we update the elements actually displayed in the slots.
+ * For example, we can start with the elements {a}, {b} and {c} assigned respectively
+ * to the slots [0], [1] and [2]. After the swipe the slots will be reassigned to {b}, {c} and {d}:
+ *
+ * start              [0]:{a} | [1]:{b} | [2]:{d}       -> gesture starts
+ * forward    [0]:{a} [1]:{b} | [2]:{c} |               -> gesture ends
+ * start              [0]:{a} | [1]:{b} | [2]:{c}       -> instant snap back to start state
+ * start              [0]:{b} | [1]:{c} | [2]:{d}       -> repaint with reassigned elements
+ *
+ * In this manner, the overall effect emulate an horizontal scroll of a list of elements.
+ *
+ * A similar mechanism is applied the left to right gesture going through the backward state.
+ *
+ * Starting slot
+ * -------------
+ *
+ * In order to operate, we need a list of slots. We retrieve them from the motionScene by adding
+ * to the slotPrefix an index number. As the starting slot may not be the first one in the scene,
+ * we also need to be able to specify a startIndex.
+ *
+ * Note that at the beginning of the Carousel, we will not populate the slots that have a lower
+ * index than startIndex, and at the end of the Carousel, we will not populate the slots that have
+ * a higher index than startIndex.
+ *
+ * @param initialSlotIndex the slot index that holds the current element
+ * @param numSlots the number of slots in the scene
+ * @param startState the name of the start state (default "start")
+ * @param backwardState the name of the backward state (default "backward")
+ * @param forwardState the name of the forward state (default "forward")
+ * @param slotPrefix the prefix used for the slots widgets in the scene (default "card")
+ * @param showSlots a debug flag to display the slots in the scene regardless if they are populated
+ * @param content the MotionCarouselScope we use to map the elements to the slots
+ */
+@OptIn(ExperimentalMotionApi::class, ExperimentalMaterialApi::class)
+@Composable
+fun MotionCarousel(
+    motionScene: MotionScene,
+    initialSlotIndex: Int,
+    numSlots: Int,
+    startState: String = "start",
+    backwardState: String = "backward",
+    forwardState: String = "forward",
```

Comment:
These seem to be transitions.
By convention we call ConstraintSets state; 
Should they be forwardTransition and backwardTransition ?